### PR TITLE
[SPARK-45424][SQL] Fix TimestampFormatter return optional parse results when only prefix match

### PR DESCRIPTION
**Pull request by Hisoka-X**
### What changes were proposed in this pull request?
When use custom pattern to parse timestamp, if there have matched prefix, not matched all. The `Iso8601TimestampFormatter::parseOptional` and `Iso8601TimestampFormatter::parseWithoutTimeZoneOptional` should not return not empty result.
eg: pattern = `yyyy-MM-dd HH:mm:ss`, value = `9999-12-31 23:59:59.999`. If fact, `yyyy-MM-dd HH:mm:ss` can parse `9999-12-31 23:59:59`  normally, but value have suffix `.999`. so we can't return not empty result.
This bug will affect inference the schema in CSV/JSON.

### Why are the changes needed?
Fix inference the schema bug.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
add new test.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Comment by andygrove** [comment]
Thanks @Hisoka-X. I tested this out, but the behavior is still different from 3.4.0. I don't think we can just use the length of the format to verify matches, because some formats have optional components like `[.SSS][XXX]`.

Using the same test file from the issue:

```csv
2884-06-24T02:45:51.138
2884-06-24T02:45:51.138
2884-06-24T02:45:51.138
```

## 3.4.0

Infers column as `timestamp`.

```
scala> val df = spark.read.option("timestampFormat", "yyyy-MM-dd'T'HH:mm:ss[.SSS][XXX]").option("inferSchema", true).csv("/tmp/timestamps.csv")
df: org.apache.spark.sql.DataFrame = [_c0: timestamp]
```

## This PR

Infers column as `string`.

```
scala> val df = spark.read.option("timestampFormat", "yyyy-MM-dd'T'HH:mm:ss[.SSS][XXX]").option("inferSchema", false).csv("/tmp/timestamps.csv")
val df: org.apache.spark.sql.DataFrame = [_c0: string]
```

**Review by andygrove — APPROVED** [review]
LGTM

**Comment by beliefer** [comment]
cc @MaxGekk

**Review by MaxGekk — COMMENTED** [review]
@Hisoka-X How about this changes https://github.com/apache/spark/commit/3192bbd29585607d43d0819c6c2d3ac00180261a . Do they suffer from the same issue?

File: sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/TimestampFormatter.scala [hunk 1]
```diff
@@ -196,8 +197,9 @@ class Iso8601TimestampFormatter(
 
   override def parseWithoutTimeZoneOptional(s: String, allowTimeZone: Boolean): Option[Long] = {
     try {
-      val parsed = formatter.parseUnresolved(s, new ParsePosition(0))
-      if (parsed != null) {
+      val parsePosition = new ParsePosition(0)
+      val parsed = formatter.parseUnresolved(s, parsePosition)
+      if (parsed != null && s.length == parsePosition.getIndex) {
```
Comment (MaxGekk) [inline sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/TimestampFormatter.scala:202]: Could you add a check/test for the changes too, please.

**Review by MaxGekk — CHANGES_REQUESTED** [review]

File: sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/TimestampFormatter.scala (hunk 1 above)
Comment (MaxGekk) [reply sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/TimestampFormatter.scala:202]: I put a breakpoint there, and the test doesn't stop on it. Only at `parseOptional()`

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master/3.5.
Thank you, @Hisoka-X and @andygrove for review.

**Pull request closed without merging**